cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -269,30 +285,406 @@ func copyFromGlobHelper(src, dst, globEx string) error {
+		return true
+	}
+	if r.UbuntuData.MountState != partitionMounted {
+		return true
+	}
+
+	// TODO: should we also check MountLocation too?
+
+	return false
+}
+
+func (m *stateMachine) diskOpts() *disks.Options {
+	if m.isEncryptedDev {
+		return &disks.Options{
+			IsDecryptedDevice: true,
+		}
+	}
+	return nil
+}
+
+func (m *stateMachine) verifyMountPoint(dir, name string) error {
+	matches, err := m.disk.MountPointIsFromDisk(dir, m.diskOpts())
+	if err != nil {
 		return err
 	}
+	if !matches {
+		return fmt.Errorf("cannot validate mount: %s mountpoint target %s is expected to be from disk %s but is not", name, dir, m.disk.Dev())
+	}
+	return nil
+}
 
-	// get the disk that we mounted the ubuntu-seed partition from as a
-	// reference point for future mounts
-	disk, err := disks.DiskFromMountPoint(boot.InitramfsUbuntuSeedDir, nil)
+func (m *stateMachine) setFindState(part string, err error, logNotFoundErr bool) error {
+	if err == nil {
+		// device was found
+		m.degradedState.partition(part).FindState = partitionFound
+		return nil
+	}
+	if _, ok := err.(disks.FilesystemLabelNotFoundError); ok {
+		// explicit error that the device was not found
+		m.degradedState.partition(part).FindState = partitionNotFound
+		if logNotFoundErr {
+			m.degradedState.LogErrorf("cannot find %v partition on disk %s", part, m.disk.Dev())
+		}
+		return nil
+	}
+	// the error is not "not-found", so we have a real error
+	m.degradedState.partition(part).FindState = partitionErrFinding
+	m.degradedState.LogErrorf("error finding %v partition on disk %s: %v", part, m.disk.Dev(), err)
+	return nil
+}
+
+func (m *stateMachine) setMountState(part, where string, err error) error {
 	if err != nil {
+		m.degradedState.LogErrorf("cannot mount %v: %v", part, err)
+		m.degradedState.partition(part).MountState = partitionErrMounting
+		return nil
+	}
+
+	m.degradedState.partition(part).MountState = partitionMounted
+	m.degradedState.partition(part).MountLocation = where
+
+	if err := m.verifyMountPoint(where, part); err != nil {
+		m.degradedState.LogErrorf("cannot verify %s mount point at %v: %v",
+			part, where, err)
 		return err
 	}
+	return nil
+}
+
+func (m *stateMachine) setUnlockStateWithRunKey(partName string, unlockRes secboot.UnlockResult, err error) error {
+	part := m.degradedState.partition(partName)
+	// save the device if we found it from secboot
+	if unlockRes.Device != "" {
+		part.FindState = partitionFound
+		part.Device = unlockRes.Device
+	} else {
+		part.FindState = partitionNotFound
+	}
+
+	if err != nil {
+		// create different error message for encrypted vs unencrypted
+		if unlockRes.IsDecryptedDevice {
+			devStr := partName
+			if unlockRes.Device != "" {
+				devStr += fmt.Sprintf(" (device %s)", unlockRes.Device)
+			}
+			m.degradedState.LogErrorf("cannot unlock encrypted %s with sealed run key: %v", devStr, err)
+		} else {
+			m.degradedState.LogErrorf("failed to find %s partition for mounting host data: %v", part, err)
```

Comment:
s/failed to/cannot/ ?

also we can't distinguish atm NotFound and ErrFinding because the Unlock function lose information